File: Roadmap/27 - SOLID OCP/python/nlarrea.py
"""
 * EJERCICIO:
 * Explora el "Principio SOLID Abierto-Cerrado (Open-Close Principle, OCP)" 
 * y crea un ejemplo simple donde se muestre su funcionamiento
 * de forma correcta e incorrecta.
 *
 * DIFICULTAD EXTRA (opcional):
 * Desarrolla una calculadora que necesita realizar diversas operaciones matemáticas. 
 * Requisitos:
 * - Debes diseñar un sistema que permita agregar nuevas operaciones utilizando el OCP.
 * Instrucciones:
 * 1. Implementa las operaciones de suma, resta, multiplicación y división.
 * 2. Comprueba que el sistema funciona.
 * 3. Agrega una quinta operación para calcular potencias.
 * 4. Comprueba que se cumple el OCP.
"""

# Incorrecto


class GeometricForm:
    def draw_square(self):
        print("Drawing a square")

    def draw_circle(self):
        print("Drawing a circle")


# Correcto


class GeometricForm:
    def draw(self):
        pass


class Square(GeometricForm):
    def draw(self):
        print("Drawing a square!")


class Circle(GeometricForm):
    def draw(self):
        print("Drawing a circle!")


"""
 * DIFICULTAD EXTRA (opcional):
 * Desarrolla una calculadora que necesita realizar diversas operaciones matemáticas. 
 * Requisitos:
 * - Debes diseñar un sistema que permita agregar nuevas operaciones utilizando el OCP.
 * Instrucciones:
 * 1. Implementa las operaciones de suma, resta, multiplicación y división.
 * 2. Comprueba que el sistema funciona.
 * 3. Agrega una quinta operación para calcular potencias.
 * 4. Comprueba que se cumple el OCP.
"""


# Import Abstract Base Class and abstractmethod decorator
from abc import ABC, abstractmethod


# Create the abstract class -> avoid's creating instances of this class, but
# still allows to inherit from it
class Operation(ABC):
    @abstractmethod
    def calculate(self, num_one, num_two):
        pass


# Create the different operations


class Sum(Operation):
    def calculate(self, num_one, num_two):
        return num_one + num_two


class Subtraction(Operation):
    def calculate(self, num_one, num_two):
        return num_one - num_two


class Multiplication(Operation):
    def calculate(self, num_one, num_two):
        return num_one * num_two


class Division(Operation):
    def calculate(self, num_one, num_two):
        return num_one / num_two


# Create the Calculator class
class Calculator:
    def __init__(self) -> None:
        self.operations = dict()

    def set_operation(self, name: str, operation: object):
        """Allow adding new operations to the current calculator instance."""

        if not isinstance(name, str):
            raise ValueError("The operation name must be an string!")
        self.operations[name] = operation

    def calculate(self, name: str, num_one, num_two):
        """Checks if the given operation name exists and returns the value of
        the operation."""

        if name not in self.operations:
            raise ValueError(f"The operation '{name}' doesn't exist!")
        return self.operations[name].calculate(num_one, num_two)


# Create the Calculator's instance and set different operations
calculator = Calculator()
calculator.set_operation("sum", Sum())
calculator.set_operation("subtract", Subtraction())
calculator.set_operation("multiply", Multiplication())
calculator.set_operation("divide", Division())

# Check if operations work
print(calculator.calculate("sum", 10, 2))  # 12
print(calculator.calculate("subtract", 10, 2))  # 8
print(calculator.calculate("multiply", 10, 2))  # 20
print(calculator.calculate("divide", 10, 2))  # 5.0
# print(calculator.calculate("pow", 10, 2))  # The operation 'pow' doesn't exist


# Create the new Pow operation
class Pow(Operation):
    def calculate(self, num_one, num_two):
        return num_one**num_two


# Set it to the calculator
calculator.set_operation("pow", Pow())

# Check if operation works
print(calculator.calculate("pow", 10, 2))  # 100

File: Roadmap/27 - SOLID OCP/javascript/nlarrea.js
/*
 * EJERCICIO:
 * Explora el "Principio SOLID Abierto-Cerrado (Open-Close Principle, OCP)" 
 * y crea un ejemplo simple donde se muestre su funcionamiento
 * de forma correcta e incorrecta.
 *
 * DIFICULTAD EXTRA (opcional):
 * Desarrolla una calculadora que necesita realizar diversas operaciones matemáticas. 
 * Requisitos:
 * - Debes diseñar un sistema que permita agregar nuevas operaciones utilizando el OCP.
 * Instrucciones:
 * 1. Implementa las operaciones de suma, resta, multiplicación y división.
 * 2. Comprueba que el sistema funciona.
 * 3. Agrega una quinta operación para calcular potencias.
 * 4. Comprueba que se cumple el OCP.
 */


// Incorrect

class BadGeometricForm {
    drawSquare() {
        console.log('Drawing a square');
    }

    drawCircle() {
        console.log('Drawing a circle');
    }
}


// Correct

class GeometricForm {
    constructor() {
        if (this.constructor === GeometricForm) {
            throw new Error('You can\'t create an instance from GeometricForm class!')
        }
    }

    draw() {
        throw new Error('The method "draw" must be implemented!')
    }
}


class Square extends GeometricForm {
    draw() {
        console.log('Drawing a square!');
    }
}


class Circle extends GeometricForm {
    draw() {
        console.log('Drawing a circle!');
    }
}


const circle = new Circle();
circle.draw();  // Drawing a circle!


/*
 * DIFICULTAD EXTRA (opcional):
 * Desarrolla una calculadora que necesita realizar diversas operaciones matemáticas. 
 * Requisitos:
 * - Debes diseñar un sistema que permita agregar nuevas operaciones utilizando el OCP.
 * Instrucciones:
 * 1. Implementa las operaciones de suma, resta, multiplicación y división.
 * 2. Comprueba que el sistema funciona.
 * 3. Agrega una quinta operación para calcular potencias.
 * 4. Comprueba que se cumple el OCP.
 */


/* Create an abstract class -> avoid creating instances of this class, but
still allows inherit from it */
class Operation {
    constructor() {
        if (this.constructor === Operation) {
            throw new Error('The class Operation can not be instantiated!');
        }
    }

    calculate(numOne, numTwo) {
        throw new Error('The method calculate() must be implemented!');
    }
}


// Create the different operations


class Sum extends Operation {
    calculate(numOne, numTwo) {
        return numOne + numTwo;
    }
}


class Subtraction extends Operation {
    calculate(numOne, numTwo) {
        return numOne - numTwo;
    }
}


class Multiplication extends Operation {
    calculate(numOne, numTwo) {
        return numOne * numTwo;
    }
}


class Division extends Operation {
    calculate(numOne, numTwo) {
        return numOne / numTwo;
    }
}


// Create the Calculator class


class Calculator {
    constructor() {
        this.operations = {};
    }

    /**
     * Allows setting a new operation to the Calculator.
     * @param {String} name The name of the operation to set
     * @param {Operation} operation The operation that must be implemented
     */
    setOperation(name, operation) {
        this.operations[name] = operation;
    }

    /**
     * @param {String} name The name of the operation to use
     * @param {Number} numOne The first number to use by the operation
     * @param {Number} numTwo The second number to use by the operation
     * @returns {Number} The result of the implemented operation
     */
    calculate(name, numOne, numTwo) {
        if (!Object.keys(this.operations).includes(name)) {
            throw new Error(`The operation "${name}" does not exist!`)
        }
        return this.operations[name].calculate(numOne, numTwo);
    }
}


// Create the calculator instance and add the different operations
const calculator = new Calculator();
calculator.setOperation('sum', new Sum());
calculator.setOperation('subtract', new Subtraction());
calculator.setOperation('multiply', new Multiplication());
calculator.setOperation('divide', new Division());

// Check if the current operations work
console.log(calculator.calculate('sum', 10, 2));  // 12
console.log(calculator.calculate('subtract', 10, 2));  // 8
console.log(calculator.calculate('multiply', 10, 2));  // 20
console.log(calculator.calculate('divide', 10, 2));  // 5
// console.log(calculator.calculate('pow', 10, 2));  // Error -> The operation "pow" does not exist!


// Create the Pow operation


class Pow extends Operation {
    calculate(numOne, numTwo) {
        return numOne ** numTwo;
    }
}


// Add the Pow operation to the calculator and check if it works
calculator.setOperation('pow', new Pow());
console.log(calculator.calculate('pow', 10, 2));  // 100